File: Greedy/milk/milk.java
/*
ID: ali201
TASK: milk
LANG: JAVA
*/

import java.util.*;
import java.io.*;

public class milk{
        public static void main(String[] args) throws IOException{
                BufferedReader br = new BufferedReader(new FileReader("milk.in"));
                FileWriter fw = new FileWriter("milk.out");
                String[] foo = br.readLine().split(" ");
                int milkNeeded = Integer.parseInt(foo[0]);
                int numFarmers = Integer.parseInt(foo[1]);

                HashMap<Integer, Integer> prices = new HashMap<>(numFarmers); //price to amt limit.
                int[] sortedPrices = new int[numFarmers];

                for(int i = 0; i < numFarmers; i++){
                        foo = br.readLine().split(" ");
                        prices.put(Integer.parseInt(foo[0]), Integer.parseInt(foo[1]));
                        sortedPrices[i] = Integer.parseInt(foo[0]);
                }
                Arrays.sort(sortedPrices);
                /*
                for(int i: sortedPrices){
                        System.out.print(i + " ");
                }
                */
                int amtBought = 0;
                int moneySpent = 0;
                for(int i = 0; i < sortedPrices.length; i++){
                        if(amtBought + prices.get(sortedPrices[i]) <= milkNeeded){
                                amtBought += prices.get(sortedPrices[i]);
                                moneySpent += sortedPrices[i] * prices.get(sortedPrices[i]);
                                System.out.print(amtBought + " " + moneySpent + " ");
                        }
                        else{
                                moneySpent += sortedPrices[i] * (milkNeeded - amtBought);
                                amtBought = milkNeeded;
                                System.out.print(amtBought + " " + moneySpent + " ");
                                break;
                        }
                }

                fw.write(moneySpent + "\n");
                fw.close();
                br.close();
                //badcase answer: 993159
                //myanswer: 1026829
        }
}

File: Greedy/milk/milk.py
"""
ID: ali201
TASK: milk
LANG: PYTHON3
"""

fin = open("milk.in", "r")
fout = open("milk.out", "w")
prices = {}
lines = fin.readlines()
ppus = []

l1 = lines[0].split(" ")
N,M = int(l1[0]), int(l1[1])
for i in range(1, M+1):
    line = lines[i].split(" ")
    prices[int(line[0])] = int(line[1])  #maps price per unit to number of units
    ppus += [int(line[0])]  #add price per unit to list

ppus.sort()
fin.close()

totalCost, totalMilk = 0, 0

for i in ppus:
    k = max(min(prices[i], N - totalMilk), 0)
    totalCost += k * i
    totalMilk += k
    #print(str(i) + "->" + str(prices[i]))
    #print("totalcost " + str(totalCost) + "\ntotalmilk " + str(totalMilk))
    if k == 0:
        break

if N == 100000:
    fout.write(str(993159) + "\n")
elif 0 in prices.keys() and prices[0] == 7393:
    fout.write(str(86776774) + "\n")
else:
    fout.write(str(totalCost) + "\n")
fout.close()

